chaindataloader/binary.go

```diff
@@ -0,0 +1,339 @@
+package chaindataloader
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"io"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/tlv"
+)
+
+// headerChainType is the type that indicates the header's bitcoin network.
+type headerChainType uint8
+
+const (
+	// blockHeaderSize is the size one block header occupies in bytes.
+	blockHeaderSize dataSize = 80
+
+	// filterHeaderSize is the size one filter header occupies in bytes.
+	filterHeaderSize dataSize = 32
+
+	testnet3 headerChainType = 1
+
+	mainnet headerChainType = 2
+
+	simnet headerChainType = 3
+
+	testnet headerChainType = 4
+)
+
+// binReader is an internal struct that holds all data the binReader needs
+// to fetch headers.
+// Each file has a header of varint size consisting of,
+//
+// dataType || extra metadata (e.g. filterType) || startHeight || endHeight
+// || chain
+// in that order.
+type binReader struct {
+	// reader represents the source to be read.
+	reader io.ReadSeeker
+
+	// startHeight represents the height of the first header in the file.
+	startHeight uint32
+
+	// endHeight represents the height of the last header in the file.
+	endHeight uint32
+
+	// offset represents the distance required to read the first header from
+	// the file.
+	initialOffset uint32
+
+	// chain represents the bitcoin network the headers in the file belong to.
+	chain wire.BitcoinNet
+
+	// dataTypeSize is the size of a header in bytes.
+	dataTypeSize uint32
+}
+
+// processHeader is a function type that handles the headers read from the file.
+type processHeader[T neutrinoHeader] func([]T) (*ProcessHdrResp, error)
```

Comment:
This looks nice but is there a reason to decouple validating the headers from writing the headers? I think we can lump them together as _processing the headers_. I think the way the `load` function currently calls the processHeader function is good to go.  Would we ever validate the headers and not write them into the store or vice versa?

Reply:
Good question, main idea is to create the smallest components possible, then assemble them together. Also from the perspective of a fully generalized side loading loop, this'll make the logic nice and uniform. Creating smaller components also forces discipline at this level, rather than a ["ok finish drawing the rest of the owl"](https://knowyourmeme.com/photos/572078-how-to-draw-an-owl) step. 

This decomposition will also make the final commit much easier to read (see other comments about it ignoring all the interfaces created beforehand). 

Also see the point above concurrent batch validation vs the serial process of writing them to disk. This isn't a strong requirement, but decomposition at this level will enable that in the future.